python/pyspark/context.py

```diff
@@ -112,6 +112,18 @@ def __init__(self, master=None, appName=None, sparkHome=None, pyFiles=None,
         ValueError:...
         """
         self._callsite = first_spark_call() or CallSite(None, None, None)
+        if gateway is not None and gateway.gateway_parameters.auth_token is None:
+            if conf and conf.get("spark.python.allowInsecurePy4j", "false") == "true":
```

Comment:
This requires the config to be set in the `SparkConf` object, instead of allowing it to be in a config file, right? I don't think the config file has been read at this point in the code.

The issue I see with that is that it would require code changes to make this work, instead of being simply something to add to a config file in apps that currently pass in insecure gateways.

(But then, given that the config file is probably read by the JVM, you'd have a chicken-and-egg problem, so maybe this is ok...)

Reply:
Would it be better if we just control this via environment variable instead? The gateway will already be in driver side so we could simply set it.

Reply:
hmm good point about not reading a file.  I can switch to an environment variable -- is that really better for a use case like zeppelin?  neither solution seems ideal, but I would really like to keep some check here that forces users to opt-in.

Reply:
I think either way is possible but I assume it's easier to handle environment variable instead? No strong opinion on this ..

Reply:
I thought about this more, and I realized that if we're requiring a code change for the end user, it sort of defeats the entire purpose of this change (eg. users need this because they can't upgrade zeppelin).  One thing I could do is delay this check until we get the full spark conf back from java:

https://github.com/apache/spark/blob/5ed89ceaf367590f79401abbf9ff7fc66507fe4e/python/pyspark/context.py#L169-L171

then the check could be based on the real conf, which includes anything set in a file

Reply:
There's no guarantee that the context initialized at that point will read any configuration file... the config file is processed by `SparkSubmit`, which as far as I can tell is not (necessarily) used in this code path.

Sounds like the env variable would be an ok compromise; not sure if it's easy to set those through something like Zeppelin, though.

Reply:
ugh, yeah I didn't think this through.  @HyukjinKwon I think you might know the zeppellin use case best -- any opinion what would be easier for the end user to change, without having to modify the code of zeppelin itself?

Reply:
I could also allow both env or conf ...

Reply:
I just happened to know a bit due to a bit of related work lately .. not an expert but both indeed are pretty much feasible, and IMHO env is easier and cleaner. 

cc @zjffdu for sure.

The argument here is basically ... the config to be set in the SparkConf object, instead of allowing it to be in a config file VS making this an env.

This conf or env might have to be set in Zeppelin side to allow insecure connection if I am not mistaken. We're not sure which one is preferred. To me, env looks preferred IMHO.

Reply:
I forget to cc one who's committer at both projects :-) @felixcheung

Reply:
If I recall, I don't think we can set SparkConf, so to me env var is better.
of course @zjffdu should know best at this point.

Reply:
scratch that.

Zeppelin is setting SparkConf before creating the SparkContext, here https://github.com/apache/zeppelin/blob/master/spark/interpreter/src/main/java/org/apache/zeppelin/spark/NewSparkInterpreter.java#L97 so I don't think it matter. either is fine and perhaps SparkConf is more consistent with other config.

Reply:
wait, lets be clear on the end goal here.  The point is for an end user to be able to enable the opt-in *without* them changing the code in zeppelin at all.  If we're going to change the code in zeppelin, it should just be changed to do the right thing and create a secure gateway (as zeppelin already has changed in master, and I think even v0.8.0 now that I look more closely).

so looking at an old version of zeppelin, eg:

https://github.com/apache/zeppelin/blob/v0.7.3/spark/src/main/java/org/apache/zeppelin/spark/PySparkInterpreter.java#L205-L229

If I'm reading that correctly, it looks like environment variables the user has set when starting zeppelin will get passed through to the python command (I think that is what `EnvironmentUtils.getProcEnvironment()` does).  But there isn't any way for the user to add additional confs to that command.  I assume things other than zeppelin would work similarly

Reply:
I updated this to require the env variable PYSPARK_ALLOW_INSECURE_GATEWAY to be set to 1 / true